src/aws/flb_aws_util.c

```diff
@@ -927,3 +932,86 @@ flb_sds_t flb_get_s3_key(const char *format, time_t time, const char *tag,
         }
         return NULL;
 }
+
+/*
+ * Replace %f and %L in time_key_format with milliseconds and nanoseconds. And get the
+ * time format output and output length with strftime().
+ */
+size_t create_time_formatter(char **out_buf, const char *time_key_format,
+                             struct flb_time *tms)
+{
+    char millisecond_str[FLB_AWS_MILLISECOND_FORMATTER_LENGTH+1];
+    char nanosecond_str[FLB_AWS_NANOSECOND_FORMATTER_LENGTH+1];
+    char *tmp_parsed_time_str;
+    char *buf;
+    size_t out_size;
+    size_t tmp_parsed_time_str_len;
+    size_t time_key_format_len;
+    struct tm timestamp;
+    struct tm *tmp;
+
+    /*
+     * Guess the max_length needed for tmp_parsed_time_str and tmp_out_buf. The
+     * upper bound is 12*strlen(time_key_format) because the worst scenario will be only
+     * %c in time_key_format, and %c will be tansfer to 24 chars long time_format
+     * by function strftime().
+     */
+    time_key_format_len = strlen(time_key_format);
+    tmp_parsed_time_str_len = 12*time_key_format_len;
+
+    /*
+     * Use tmp_parsed_time_str to buffer time_key_format when replace %f and %F
+     * with milliseconds and nanoseconds.
+     */
+    tmp_parsed_time_str = (char *)flb_calloc(tmp_parsed_time_str_len, sizeof(char));
+    if (!tmp_parsed_time_str) {
+        flb_errno();
+        return -1;
+    }
+
+    buf = (char *)flb_calloc(tmp_parsed_time_str_len, sizeof(char));
+    if (!buf) {
+        flb_errno();
+        flb_free(tmp_parsed_time_str);
+        return -1;
+    }
+
+    /* Replace %f to millisecond, %L to nanosecond in time_key_format. */
+    snprintf(millisecond_str, FLB_AWS_MILLISECOND_FORMATTER_LENGTH+1,
+             "%" PRIu64, (uint64_t) tms->tm.tv_nsec / 1000000);
+    snprintf(nanosecond_str, FLB_AWS_NANOSECOND_FORMATTER_LENGTH+1,
+             "%" PRIu64, (uint64_t) tms->tm.tv_nsec);
+    for (int i = 0; i < time_key_format_len; i++) {
+        if (strncmp(time_key_format+i, FLB_AWS_MILLISECOND_FORMATTER, 2) == 0) {
```

Comment:
So this compares 2 chars in the string, but then `i` can advance to the final char of the string... one might fear that could lead to an issue except that strncmp will protect us here... can you explain to me why?